### PR TITLE
Add `stop-pjax-loading-with-esc` feature

### LINKED ISSUE #2118: Stop ajax loading with `esc`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/refined-github/issues/2118)

# Current

1. Click on the "Issues" tab above
2. Change your mind
3. Can't do anything, Issues keeps loading
4. When loaded, go back

# Desired

1. Click on the "Issues" tab above
2. Change your mind
3. Press `esc`, loading stops

⬆️ this is how regular pages work, but `pjax` doesn't listen to `esc`

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/refined-github/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by artusm**
1. LINKED ISSUES:
   Closes #2118

2. TEST URLS:
   [sindresorhus/refined-github](https://github.com/sindresorhus/refined-github)

3. SCREENSHOT:
   
Peek 2020-08-16 02-27

## Known Bugs
- <s>Few bugs with profile page after cancelling load on the profile page.</s> (fixed)Peek 2020-08-16 03-00
- <s>Browser forward button doesn't work correctly after cancelling load.</s> (fixed)Peek 2020-08-16 10-25

### CODE REVIEW

**Comment by fregante** [comment]
This is honestly much better than I'd have thought, I don't even see the profile issue

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+		}
+	});
+}
+
+void features.add({
+	id: __filebasename,
+	description: '',
+	screenshot: false
+}, {
+	init: onetime(init)
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:65]: Can you add some specific pages where this works? e.g. `isConversationList`, `isPR`, `isProfile`. Not the generic `isRepo` though

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	});
+
+	window.addEventListener('pjax:error', event => {
+		if (typeof event.cancelable !== 'boolean' || event.cancelable) {
+			event.preventDefault();
+		}
+	});
+
+	window.addEventListener('pjax:end', () => {
+		if (pageDetect.isUserProfile()) {
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:54]:
```suggestion
	if (pageDetect.isUserProfile()) {
		window.addEventListener('pjax:end', () => {
```

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+				}, '', location.href);
+			}
+
+			history.back();
+			progressLoader.classList.remove(progressLoaderLoadingClass);
+		}
+	});
+
+	window.addEventListener('pjax:error', event => {
+		if (typeof event.cancelable !== 'boolean' || event.cancelable) {
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:48]: When is `cancelable` not a boolean?

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+function init() {
+	const progressLoader = select('.progress-pjax-loader')!;
+
+	window.addEventListener('keydown', event => {
+		if (event.key === 'Escape' && progressLoader.classList.contains(progressLoaderLoadingClass)) {
+			const pjaxContainerId = getPjaxContainerId();
+
+			if (pjaxContainerId && history.state && '_id' in history.state) {
+				history.replaceState({
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:34]: Can you add a comment here explaining why this is needed?

**Comment by artusm** [comment]
> I don't even see the profile issue

Because I've fixed it

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+function init() {
+	const progressLoader = select('.progress-pjax-loader')!;
+
+	window.addEventListener('keydown', event => {
+		if (event.key === 'Escape' && progressLoader.classList.contains(progressLoaderLoadingClass)) {
+			const pjaxContainerId = getPjaxContainerId();
+
+			if (pjaxContainerId && history.state && '_id' in history.state) {
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:33]: It seems that `pjaxContainerId` is only needed in some cases, but `features.error` logs an error unconditionally. Maybe this should be
```suggestion
			if (history.state && '_id' in history.state) {
				const pjaxContainer = select('#js-repo-pjax-container, #js-pjax-container, #gist-pjax-container');
				if (pjaxContainer) {
					history.replaceState(.........)
				} else {
					features.error(__filebasename, 'Pjax container id not found.');
				}
			}
```

File: source/features/stop-pjax-loading-with-esc.tsx (hunk 3 above)
Comment (artusm) [reply source/features/stop-pjax-loading-with-esc.tsx:48]: I don't know. I just copy&pasted from [this](https://developer.mozilla.org/en-US/docs/Web/API/Event/cancelable) example.

File: source/features/stop-pjax-loading-with-esc.tsx (hunk 1 above)
Comment (fregante) [reply source/features/stop-pjax-loading-with-esc.tsx:65]: No, use `include`, not `exclude`

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	for (const stickyElement of select.all('.js-sticky:not(.is-stuck)')) {
+		stickyElement.removeAttribute('style');
+	}
+}
+
+function init() {
+	const progressLoader = select('.progress-pjax-loader')!;
+
+	window.addEventListener('keydown', event => {
+		if (event.key === 'Escape' && progressLoader.classList.contains(progressLoaderLoadingClass)) {
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:19]: Can you move all listeners out of `init`?

Also let's use an early return (`if (!) {return}`) to limit the indentation further

File: source/features/stop-pjax-loading-with-esc.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+		pageDetect.isRepoCommitList,
+		pageDetect.isCompare,
+		pageDetect.isGlobalSearchResults,
+		pageDetect.isLabelList,
+		pageDetect.isMilestone,
+		pageDetect.isMilestoneList,
+		pageDetect.isNewIssue,
+		pageDetect.isPRConflicts,
+		pageDetect.isQuickPR,
+		pageDetect.isEditingFile,
```
Comment (fregante) [inline source/features/stop-pjax-loading-with-esc.tsx:80]: I don't think many of these are loaded via ajax. Maybe  `include: isRepo, isGlobalSearchResults` was the right solution after all

Comment (artusm) [reply source/features/stop-pjax-loading-with-esc.tsx:80]: > I don't think many of these are loaded via ajax.

Yup, BUT these pages contains pjax-links.

File: readme.md [hunk 8]
```diff
@@ -324,6 +324,7 @@ Thanks for contributing! 🦋🙌
 - [](# "action-used-by-link") [Lets you see how others are using the current Action in the Marketplace.](https://user-images.githubusercontent.com/8360597/80250140-86d9c080-8673-11ea-9d28-f62faf9fd3d4.png)
 - [](# "improve-shortcut-help") [Shows all of Refined GitHub’s new keyboard shortcuts in the help modal.](https://user-images.githubusercontent.com/29176678/36999174-9f07d33e-20bf-11e8-83e3-b3a9908a4b5f.png) *(<kbd>?</kbd> hotkey)*
 - [](# "stop-redirecting-in-notification-bar") [Stops redirecting to notification inbox from notification bar actions while holding <kbd>Alt</kbd>.](https://user-images.githubusercontent.com/202916/80318782-c38cef80-880c-11ea-9226-72c585f42a51.png)
+- [](# "stop-pjax-loading-with-esc") [Stops pjax loading by pressing the <kbd>Esc</kbd> button.](https://user-images.githubusercontent.com/36174850/90323385-3c08ef00-df69-11ea-8c0e-c85241888a7b.gif)
```
Comment (fregante) [inline readme.md:327]: This should be moved to the section below. It’s a GitHub bug after all

File: readme.md [hunk 9]
```diff
@@ -348,6 +348,7 @@ Thanks for contributing! 🦋🙌
 - [](# "align-repository-header") [Aligns the repository header to the repository content on wide screens.](https://user-images.githubusercontent.com/1402241/86574587-587f3800-bf76-11ea-9961-5c25cdb6e357.gif)
 - [](# "linkify-user-edit-history-popup") [Linkifies the username in the edit history popup.](https://user-images.githubusercontent.com/1402241/88917988-9ebb7480-d260-11ea-8690-0a3440f1ebbc.png)
 - [](# "prevent-duplicate-pr-submission") [Avoids creating duplicate PRs when mistakenly clicking "Create pull request" more than once.](https://user-images.githubusercontent.com/16872793/89589967-e029c200-d814-11ea-962b-3ff1f6236781.gif)
+- [](# "stop-pjax-loading-with-esc") [Stops pjax loading by pressing the <kbd>Esc</kbd> button.](https://user-images.githubusercontent.com/36174850/90323385-3c08ef00-df69-11ea-8c0e-c85241888a7b.gif)
```
Comment (fregante) [inline readme.md:351]:
```suggestion
- [](# "stop-pjax-loading-with-esc") [After you click on an ajaxed link, this lets you stop loading a page by pressing the <kbd>esc</kbd> key, like the browser does for regular page loads.](https://user-images.githubusercontent.com/36174850/90323385-3c08ef00-df69-11ea-8c0e-c85241888a7b.gif)
```

**Review by fregante — APPROVED** [review]

**Comment by fregante** [comment]
We're just waiting for @sindresorhus’s review

**Review by 641210 — COMMENTED** [review]
****

**Pull request merged**